### PR TITLE
Add single partition router

### PR DESCRIPTION
**Pull request by crossoverJie**
### Motivation
image

https://pulsar.apache.org/docs/2.11.x/concepts-messaging/#routing-modes
https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/SinglePartitionMessageRouterImpl.java

Based on the code of the Java client, Implemented the `single-partition-router` mentioned in the document.

### Modifications

Add single-partition-router

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

- https://github.com/crossoverJie/pulsar-client-go/blob/add-single-partition-router/pulsar/single_partition_router_test.go
- https://github.com/crossoverJie/pulsar-client-go/blob/add-single-partition-router/pulsar/single_partition_route_bench_test.go
- https://github.com/crossoverJie/pulsar-client-go/blob/ae50c8bd6a960410069e2568ea9aeb1e4d27a625/pulsar/producer_test.go#L563

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes) Add `NewSinglePartitionRouter()` method.
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (GoDocs)

### CODE REVIEW

**Comment by shibd** [comment]
@crossoverJie Hi, thanks for your contribution. Can you fix the ci lint failed?

**Comment by crossoverJie** [comment]
> @crossoverJie Hi, thanks for your contribution. Can you fix the ci lint failed?
PTAL

**Comment by BewareMyPower** [comment]
Please rebase to master to have tests fixed.

**Comment by crossoverJie** [comment]
> Please rebase to master to have tests fixed.

Done

File: pulsar/single_partition_router.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+		if len(message.Key) != 0 {
+			// When a key is specified, use the hash of that key
+			return int(getHashingFunction(JavaStringHash)(message.Key) % numPartitions)
+		}
+		if singlePartition != nil {
+			return *singlePartition
+		}
+		mutex.Lock()
+		defer mutex.Unlock()
+		partition := r.R.Intn(int(numPartitions))
```
Comment (shibd) [inline pulsar/single_partition_router.go:44]: We need to judge again here if `singlePartition` is nil. Set it value only when it is nil.

**Review by shibd — APPROVED** [review]

File: pulsar/single_partition_router.go (hunk 1 above)
Comment (crossoverJie) [reply pulsar/single_partition_router.go:44]: Would using sync.Once be better?

Comment (shibd) [reply pulsar/single_partition_router.go:44]: Agree.

Comment (crossoverJie) [reply pulsar/single_partition_router.go]: Done

**Comment by crossoverJie** [comment]
> Thank you for your contribution! But the default router supports this feature:
> 
> https://github.com/apache/pulsar-client-go/blob/a119bab0f8598601c0eb7f0fcd97da7ab06700c7/pulsar/default_router.go#L57-L65

But when there are no `OrderdingKey` and `Key`, the user needs to specify specific partitions manually, here refers to the API of the Java SDK, where the SDK selects the partitions, keeping consistent with the implementation of the Java SDK.

https://github.com/apache/pulsar/blob/46b6dcd9571262f72f963a755c8a51fb4f66c81e/pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerStatTest.java#L153C1-L155

File: pulsar/single_partition_router.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+			// When an OrderingKey is specified, use the hash of that key
+			return int(getHashingFunction(JavaStringHash)(message.OrderingKey) % numPartitions)
+		}
+
+		if len(message.Key) != 0 {
+			// When a key is specified, use the hash of that key
+			return int(getHashingFunction(JavaStringHash)(message.Key) % numPartitions)
+		}
+		once.Do(func() {
+			if singlePartition == nil {
```
[the remark this answers is not in the document]
Comment (crossoverJie) [reply pulsar/single_partition_router.go]: Done

File: pulsar/single_partition_route_bench_test.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+package pulsar
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func BenchmarkSinglePartitionRouter(b *testing.B) {
```
[the remark this answers is not in the document]
Comment (crossoverJie) [reply pulsar/single_partition_route_bench_test.go]: Has been deleted

File: pulsar/producer_test.go [hunk 4]
```diff
@@ -560,6 +560,54 @@ func TestMessageRouter(t *testing.T) {
...
+	}
+	client, err := NewClient(ClientOptions{
+		URL: serviceURL,
+	})
+
+	assert.Nil(t, err)
+	defer client.Close()
+
+	// Only subscribe on the specific partition
+	consumer, err := client.Subscribe(ConsumerOptions{
```
[the remark this answers is not in the document]
Comment (crossoverJie) [reply pulsar/producer_test.go]: Has been deleted

File: pulsar/single_partition_router.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+import "sync"
+
+var (
+	singlePartition *int
+	once            sync.Once
+)
+
+func NewSinglePartitionRouter() func(*ProducerMessage, TopicMetadata) int {
+	return func(message *ProducerMessage, metadata TopicMetadata) int {
```
[the remark this answers is not in the document]
Comment (crossoverJie) [reply pulsar/single_partition_router.go:27]: Done

File: pulsar/producer_test.go [hunk 6]
```diff
@@ -560,6 +560,38 @@ func TestMessageRouter(t *testing.T) {
...
+	defer producer.Close()
+
+	ctx := context.Background()
+
+	ID, err := producer.Send(ctx, &ProducerMessage{
+		Payload: []byte("hello"),
+	})
+	assert.Nil(t, err)
+	assert.NotNil(t, ID)
+
```
Comment (RobertIndie) [inline pulsar/producer_test.go:593]: This test also tests the publishing logic. But it doesn't test the single partition router. We may need to add some assertions for it.

**Review by RobertIndie — CHANGES_REQUESTED** [review]

File: pulsar/producer_test.go (hunk 6 above)
Comment (crossoverJie) [reply pulsar/producer_test.go:593]: Need to expose a function `GetSinglePartition() int` that returns the current partition, then only consume messages from that single partition. 

Any other suggestions?

Comment (RobertIndie) [reply pulsar/producer_test.go:593]: We shouldn't expose the method just for testing.
You can refer to this code:
https://github.com/apache/pulsar-client-go/blob/7f91b2bcd798e1dc6ff27b10a557c8ee5440a83a/pulsar/producer_test.go#L494-L509

Create a consumer, and assert that all the messages are consumed from the same partition topic.

Comment (crossoverJie) [reply pulsar/producer_test.go]: That's a good suggestion, I have already made the changes.

**Review by RobertIndie — APPROVED** [review]

**Comment by crossoverJie** [comment]
CI run failed, I created another [PR](https://github.com/apache/pulsar-client-go/pull/1037) to fix this issue.

**Review by RobertIndie — APPROVED** [review]

**Comment by shibd** [comment]
@nodece Hi, Can you take a look at it again?

**Review by shibd — APPROVED** [review]

**Pull request merged**